File: cpp/cpp-boost/demo06a-list-threads.cpp
/*
LIST OF MULTIPLE THREADS
Version A: Using standard arrays
*/


#include <iostream>
#include <boost/thread.hpp>
using namespace std;



void doTask(int index) {
    cout << index;
}



int main() {
    const int NUM_THREADS = 5;

    boost::thread lstTh[NUM_THREADS];

    for (int i = 0; i < NUM_THREADS; ++i) {
        lstTh[i] = boost::thread(doTask, i);
    }

    for (int i = 0; i < NUM_THREADS; ++i) {
        lstTh[i].join();
    }

    cout << endl;
    return 0;
}

File: js-nodejs/demo06a-list-threads.js
/*
LIST OF MULTIPLE THREADS
VERSION A: Using standard 'Worker' class
*/

import { Worker, isMainThread, workerData } from 'worker_threads';


//---------------------------------------------
//            WORKER THREAD SECTION
//---------------------------------------------


const doTask = index => {
  console.log(index);
};

const workerFunc = () => {
  try {
    doTask(workerData);
  } catch (error) {
    console.error(error);
  }
};


//---------------------------------------------
//             MAIN THREAD SECTION
//---------------------------------------------


const mainFunc = () => {
  try {

    const NUM_THREADS = 5;
    const lstTh = new Array(NUM_THREADS);

    for (let i = 0; i < NUM_THREADS; ++i) {
      lstTh[i] = new Worker(new URL(import.meta.url), { workerData: i });
    }

    console.log('Good bye');

  } catch (error) {
    console.error(error);
  }
};


if (isMainThread) {
  mainFunc();
} else {
  workerFunc();
}
